### PR TITLE
Add initial support for Raspberry Pi CM5 on Yellow

### PR DESCRIPTION
**Pull request by sairon**
Add support for Raspberry Pi CM5 on Yellow. Most of the core work is the device tree with definition of Yellow peripherals, and U-Boot patches adding support for PCIe/NVMe for BCM2712, which is not present in the upstream U-Boot.

Yellow has been migrated to U-Boot 2024.10 which already has basic support for BCM2712. Other Raspberry Pi targets are still left 2024.01, so we will temporarily have two patchsets for RPi.

The PCIe/NVMe implementation in U-Boot is inspired by Raspberry Pi's downstream Linux implementation, and it might be worthwhile to check upstream implementation [1] once it gets finished but at the time of writing the patches, I was unable to make the controller work with changes from the v2 of the series.

The device tree for CM5 uses the same aliases of serial ports and other peripherals to match CM4 on Yellow, making CM5 a drop-in replacement of the CM4. Of course, the OS must be updated first before swapping the modules. For NVMe-only installs (i.e. those using CM4 Lite or being installed wittingly directly to NVMe) no further steps should be needed. For installations running from eMMC, the CM5 must be bootstrapped using the `rpiboot` utility, because the EEPROM bootloader doesn't support `BCM-USB-MSD` boot mode on CM5 anymore.

**Things known to be not working:**

**Factory reset in U-Boot**

Because GPIOs are handled by RP1 chipset connected over PCIe, there is no support for reading and setting GPIOs in current U-Boot. Adding at least some rudimentary driver to control GPIOs will be needed to make the standard Yellow's factory reset working.

**U-Boot USB**

USB (the dwc2 peripheral) doesn't seem to be working and can't access the ports behind the USB hub, neither can't be used for USB Mass Storage (after running `ums` command in U-Boot the USB device fails to be enumerated on host). This is also blocker for reflashing the Yellow's CM from U-Boot.

[1] https://lore.kernel.org/all/20241025124515.14066-1-svarbanov@suse.de/

## Summary by CodeRabbit

- **New Features**
	- Added NVMe as a boot target for the Raspberry Pi Compute Module 4, enhancing boot options.
	- Introduced a new driver for the Broadcom STB reset controller and a new driver for the Broadcom RESCAL reset controller.
	- Added basic support for the BCM2712 PCIe controller, improving compatibility and functionality.
	- Implemented a new command `fileenv` in U-Boot to read files into environment variables.
	- Added device tree configurations for the Compute Module 5.

- **Bug Fixes**
	- Reverted changes that caused boot hangs with USB devices on Raspberry Pi 4.
	- Addressed probing issues for the PCIe controller by introducing a `u-boot,no-probe` property.

- **Documentation**
	- Updated configuration files to reflect new device tree settings for Compute Module 5 and adjustments for Compute Module 4.

- **Chores**
	- Enhanced U-Boot configuration for Raspberry Pi 4 with new USB gadget functionalities and support for BCM2712 chipset.

### CODE REVIEW

File: buildroot-external/board/raspberrypi/yellow/uboot.config [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
+CONFIG_MMC_SDHCI_BCMSTB=y
+
+CONFIG_DM_USB_GADGET=y
+
+CONFIG_USB_GADGET=y
+CONFIG_USB_GADGET_MANUFACTURER="FSL"
+CONFIG_USB_GADGET_VENDOR_NUM=0x0525
+CONFIG_USB_GADGET_PRODUCT_NUM=0xa4a5
+CONFIG_USB_GADGET_DWC2_OTG=y
+CONFIG_USB_GADGET_DOWNLOAD=y
```
Comment (agners) [inline buildroot-external/board/raspberrypi/yellow/uboot.config:13]: Is this working? If not, maybe remove for now? :thinking:

**Review by agners — APPROVED** [review]
I did not dive into the device trees very deep, they are a bit PITA to review. I see we have a couple of workarounds and things relating to downstream/upstream differences :cry:  I hope this get fewer as time moves on. And of course, ideally the PCIe/NVMe patches should get obsolete at one time too :crossed_fingers: .

Generally LGTM, thanks :pray: !

File: buildroot-external/board/raspberrypi/yellow/uboot.config (hunk 1 above)
Comment (sairon) [reply buildroot-external/board/raspberrypi/yellow/uboot.config:13]: It works on CM4, so we should keep it enabled.

File: buildroot-external/board/raspberrypi/yellow/patches/linux/0016-ARM-dts-bcm2712-Add-device-tree-for-CM5-on-HA-Yellow.patch [hunk 2]
```diff
@@ -0,0 +1,1092 @@
...
++
++	wl_on_reg: wl_on_reg {
++		compatible = "regulator-fixed";
++		regulator-name = "wl-on-regulator";
++		regulator-min-microvolt = <3300000>;
++		regulator-max-microvolt = <3300000>;
++		pinctrl-0 = <&wl_on_pins>;
++		pinctrl-names = "default";
++
++		gpio = <&gio 28 GPIO_ACTIVE_HIGH>;
```
[the remark this answers is not in the document]
Comment (sairon) [reply buildroot-external/board/raspberrypi/yellow/patches/linux/0016-ARM-dts-bcm2712-Add-device-tree-for-CM5-on-HA-Yellow.patch:142]: Incorrect: https://www.kernel.org/doc/Documentation/devicetree/bindings/regulator/fixed-regulator.txt

File: buildroot-external/ota/rauc-hook [hunk 3]
```diff
@@ -88,6 +88,25 @@ check_grubenv() {
...
+        fi
+
+        if ! grep -q "\[cm5\]" "${BOOT_MNT}/config.txt"; then
+            echo "Adding CM5 config to config.txt"
+
+            # Remove old single device_tree config and add CM-specific ones
+            sed -i '/device_tree=bcm2711-rpi-cm4-ha-yellow.dtb/d; s/\[all\]/\[cm4\]\ndevice_tree=bcm2711-rpi-cm4-ha-yellow.dtb\n\n\[cm5\]\ndevice_tree=bcm2712-rpi-cm5-ha-yellow.dtb\n\n\[all\]/' "${BOOT_MNT}/config.txt"
+        fi
+    fi
+}
```
[the remark this answers is not in the document]
Comment (sairon) [reply buildroot-external/ota/rauc-hook:108]: This is overkill with a little benefit, I'm fine with the current migration.

File: buildroot-external/board/raspberrypi/patches/uboot/2024.10/0008-reset-reset-brcmstb-Add-Broadcom-STB-reset-controlle.patch [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
++struct brcmstb_reset_priv {
++	void __iomem *base;
++};
++
++static int brcmstb_reset_assert(struct reset_ctl *rst)
++{
++	unsigned int off = SW_INIT_BANK(rst->id) * SW_INIT_BANK_SIZE;
++	struct brcmstb_reset_priv *priv = dev_get_priv(rst->dev);
++
++	writel_relaxed(SW_INIT_BIT(rst->id), priv->base + off + SW_INIT_SET);
```
[the remark this answers is not in the document]
Comment (sairon) [reply buildroot-external/board/raspberrypi/patches/uboot/2024.10/0008-reset-reset-brcmstb-Add-Broadcom-STB-reset-controlle.patch:97]: The driver will be used only for a single platform which has `writel_relaxed` available, so no need to divert from the code ported from upstream.

File: buildroot-external/board/raspberrypi/patches/uboot/2024.10/0008-reset-reset-brcmstb-Add-Broadcom-STB-reset-controlle.patch [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
++	return 0;
++}
++
++U_BOOT_DRIVER(brcmstb_reset) = {
++	.name = "brcmstb-reset",
++	.id = UCLASS_RESET,
++	.of_match = brcmstb_reset_ids,
++	.ops = &brcmstb_reset_reset_ops,
++	.probe = brcmstb_reset_probe,
++	.priv_auto	= sizeof(struct brcmstb_reset_priv),
```
[the remark this answers is not in the document]
Comment (sairon) [reply buildroot-external/board/raspberrypi/patches/uboot/2024.10/0008-reset-reset-brcmstb-Add-Broadcom-STB-reset-controlle.patch:144]: This is not true anymore for quite a while.

File: buildroot-external/board/raspberrypi/patches/uboot/2024.10/0009-reset-reset-brcmstb-rescal-Add-Broadcom-RESCAL-reset.patch [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+@@ -69,6 +69,12 @@ config RESET_BRCMSTB
+ 	  This enables the reset controller driver for Broadcom STB SoCs using
+ 	  a SUN_TOP_CTRL_SW_INIT style controller.
+ 
++config RESET_BRCMSTB_RESCAL
++	tristate "Broadcom STB RESCAL reset controller"
++	help
++	  This enables the RESCAL reset controller found on BCM2712 and some other
++	  Broadcom STB SoCs (BCM7216).
++
```
[the remark this answers is not in the document]
Comment (sairon) [reply buildroot-external/board/raspberrypi/patches/uboot/2024.10/0009-reset-reset-brcmstb-rescal-Add-Broadcom-RESCAL-reset.patch:36]: Good point, will fix.

File: buildroot-external/board/raspberrypi/yellow/patches/linux/0016-ARM-dts-bcm2712-Add-device-tree-for-CM5-on-HA-Yellow.patch [hunk 7]
```diff
@@ -0,0 +1,1092 @@
...
++	clock-frequency = <400000>;
++	pinctrl-0 = <&i2c3_m4_agpio0_pins>;
++	pinctrl-names = "default";
++};
++
++/ {
++	chosen: chosen {
++		bootargs = "reboot=w coherent_pool=1M 8250.nr_uarts=1 pci=pcie_bus_safe";
++		stdout-path = "serial10:115200n8";
++	};
```
[the remark this answers is not in the document]
Comment (sairon) [reply buildroot-external/board/raspberrypi/yellow/patches/linux/0016-ARM-dts-bcm2712-Add-device-tree-for-CM5-on-HA-Yellow.patch:461]: Ignoring this, keeping the same approach as in CM4-Yellow DTS.

File: buildroot-external/board/raspberrypi/yellow/patches/linux/0016-ARM-dts-bcm2712-Add-device-tree-for-CM5-on-HA-Yellow.patch [hunk 8]
```diff
@@ -0,0 +1,1092 @@
...
++			default-state = "off";
++			linux,default-trigger = "none";
++		};
++
++		led_act: led-act {
++			label = "ACT";
++			gpios = <&gio_aon 9 GPIO_ACTIVE_LOW>;
++			default-state = "off";
++			linux,default-trigger = "mmc0";
++		};
```
[the remark this answers is not in the document]
Comment (sairon) [reply buildroot-external/board/raspberrypi/yellow/patches/linux/0016-ARM-dts-bcm2712-Add-device-tree-for-CM5-on-HA-Yellow.patch:106]: Ignoring this, keeping the same approach as in CM4-Yellow DTS.

File: buildroot-external/board/raspberrypi/patches/uboot/2024.10/0010-pci-pcie-brcmstb-Add-basic-support-for-BCM2712-PCIe.patch [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
++static int bcrm_pcie_perst_set_2712(struct brcm_pcie *pcie, u32 val)
++{
++	/* Perst bit has moved and assert value is 0 */
++	if (val)
++		clrbits_le32(pcie->base + PCIE_MISC_PCIE_CTRL, PCIE_MISC_PCIE_CTRL_PCIE_PERSTB_MASK);
++	else
++		setbits_le32(pcie->base + PCIE_MISC_PCIE_CTRL, PCIE_MISC_PCIE_CTRL_PCIE_PERSTB_MASK);
++
++	return 0;
++}
```
[the remark this answers is not in the document]
Comment (sairon) [reply buildroot-external/board/raspberrypi/patches/uboot/2024.10/0010-pci-pcie-brcmstb-Add-basic-support-for-BCM2712-PCIe.patch:226]: Will fix.

File: buildroot-external/board/raspberrypi/patches/uboot/2024.10/0010-pci-pcie-brcmstb-Add-basic-support-for-BCM2712-PCIe.patch [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+ 
++static int bcrm_pcie_perst_set_generic(struct brcm_pcie *pcie, u32 val)
++{
++	if (val)
++		setbits_le32(pcie->base + PCIE_RGR1_SW_INIT_1, RGR1_SW_INIT_1_PERST_MASK);
++	else
++		clrbits_le32(pcie->base + PCIE_RGR1_SW_INIT_1, RGR1_SW_INIT_1_PERST_MASK);
++
++	return 0;
++}
```
[the remark this answers is not in the document]
Comment (sairon) [reply buildroot-external/board/raspberrypi/patches/uboot/2024.10/0010-pci-pcie-brcmstb-Add-basic-support-for-BCM2712-PCIe.patch:190]: Will fix.

File: buildroot-external/board/raspberrypi/patches/uboot/2024.10/0005-nvme-Use-pointer-for-CPU-addressed-buffers.patch [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+ 	u64 *prp_pool;
+ 	int length = total_len;
+ 	int i, nprps;
+@@ -63,10 +63,10 @@ static int nvme_setup_prps(struct nvme_dev *dev, u64 *prp2,
+ 	}
+ 
+ 	if (length)
+-		dma_addr += (page_size - offset);
++		buffer += (page_size - offset);
+
```
[the remark this answers is not in the document]
Comment (sairon) [reply buildroot-external/board/raspberrypi/patches/uboot/2024.10/0005-nvme-Use-pointer-for-CPU-addressed-buffers.patch:41]: Reading the commit message for the patch, I think that using void pointers is the point of the whole thing. Also, it's not a new code, so ignoring this.

**Pull request merged**